File: mcp-tools/src/tools/curriculumValidator.ts
export const CURRICULUM_VALIDATOR_TOOL = {
  name: "validate_question",
  description: "Validates a maths question against UK National Curriculum Year 3 standards. Checks multiplication tables (2-10), division (max 100), place value (to 1000), and word problem requirements.",
  inputSchema: {
    type: "object",
    properties: {
      question: {
        type: "string",
        description: "The question text to validate",
      },
      topic: {
        type: "string",
        description: "The mathematical topic (e.g., multiplication, division, addition, place_value, word_problems)",
      },
    },
    required: ["question", "topic"],
  },
};

interface ValidationResult {
  is_valid: boolean;
  issues: string[];
  suggestions: string[];
  curriculum_alignment: {
    topic_appropriate: boolean;
    difficulty_level: string;
    year_group_match: boolean;
  };
}

// UK Year 3 curriculum standards
const YEAR_3_STANDARDS = {
  multiplication: {
    tables: [2, 3, 4, 5, 8, 10], // Year 3 focuses on these tables
    max_product: 100,
  },
  division: {
    max_dividend: 100,
    tables: [2, 3, 4, 5, 8, 10],
  },
  addition: {
    max_sum: 1000,
    mental_max: 100,
  },
  subtraction: {
    max_value: 1000,
    mental_max: 100,
  },
  place_value: {
    max_number: 1000,
    concepts: ["hundreds", "tens", "ones", "compare", "order"],
  },
  fractions: {
    denominators: [2, 3, 4, 8], // Unit fractions focus
    types: ["unit fractions", "equivalent", "comparing"],
  },
  word_problems: {
    max_steps: 2,
    operations: ["addition", "subtraction", "multiplication", "division"],
  },
};

function extractNumbers(text: string): number[] {
  const matches = text.match(/\d+/g);
  return matches ? matches.map(Number) : [];
}

function detectOperations(text: string): string[] {
  const operations: string[] = [];
  const lowerText = text.toLowerCase();

  if (/(\+|add|plus|sum|total|altogether|more than|increase)/.test(lowerText)) {
    operations.push("addition");
  }
  if (/(-|subtract|minus|take away|difference|less than|decrease|left|remain)/.test(lowerText)) {
    operations.push("subtraction");
  }
  if (/(×|x|multiply|times|groups of|lots of|each|per)/.test(lowerText)) {
    operations.push("multiplication");
  }
  if (/(÷|\/|divide|share|split|equally|each get)/.test(lowerText)) {
    operations.push("division");
  }

  return operations;
}

export function validateQuestion(question: string, topic: string): ValidationResult {
  const issues: string[] = [];
  const suggestions: string[] = [];
  const numbers = extractNumbers(question);
  const operations = detectOperations(question);

  let topicAppropriate = true;
  let difficultyLevel = "core";
  let yearGroupMatch = true;

  const standards = YEAR_3_STANDARDS[topic as keyof typeof YEAR_3_STANDARDS];

  if (!standards) {
    issues.push(`Unknown topic: ${topic}`);
    topicAppropriate = false;
  }

  switch (topic) {
    case "multiplication": {
      const multStandards = YEAR_3_STANDARDS.multiplication;
      
      // Check if numbers are within Year 3 times tables
      for (const num of numbers) {
        if (num > 10 && !multStandards.tables.includes(num)) {
          // Check if it's a valid product
          const isValidProduct = multStandards.tables.some(
            (t) => num % t === 0 && num / t <= 10
          );
          if (!isValidProduct && num > multStandards.max_product) {
            issues.push(`Number ${num} exceeds Year 3 multiplication range`);
            yearGroupMatch = false;
          }
        }
      }

      // Check for two-factor multiplication
      if (numbers.length >= 2) {
        const [a, b] = numbers;
        if (a > 12 || b > 12) {
          suggestions.push("Consider using factors from 2-10 times tables");
          difficultyLevel = "challenge";
        }
      }
      break;
    }

    case "division": {
      const divStandards = YEAR_3_STANDARDS.division;
      const maxNum = Math.max(...numbers, 0);

      if (maxNum > divStandards.max_dividend) {
        issues.push(`Dividend ${maxNum} exceeds Year 3 maximum of ${divStandards.max_dividend}`);
        yearGroupMatch = false;
      }

      // Check for division by appropriate numbers
      if (numbers.length >= 2) {
        const divisor = numbers[1];
        if (!divStandards.tables.includes(divisor)) {
          suggestions.push(`Consider using a divisor from Year 3 tables: ${divStandards.tables.join(", ")}`);
        }
      }
      break;
    }

    case "addition":
    case "subtraction": {
      const opStandards = YEAR_3_STANDARDS[topic];
      const maxNum = Math.max(...numbers, 0);

      if (maxNum > opStandards.max_sum || maxNum > opStandards.max_value) {
        issues.push(`Numbers exceed Year 3 ${topic} range (max ${opStandards.max_sum || opStandards.max_value})`);
        yearGroupMatch = false;
      }

      if (maxNum > opStandards.mental_max) {
        difficultyLevel = "challenge";
        suggestions.push("This may require written methods rather than mental calculation");
      }
      break;
    }

    case "place_value": {
      const pvStandards = YEAR_3_STANDARDS.place_value;
      const maxNum = Math.max(...numbers, 0);

      if (maxNum > pvStandards.max_number) {
        issues.push(`Number ${maxNum} exceeds Year 3 place value range (max 1000)`);
        yearGroupMatch = false;
      }

      // Check for appropriate concepts
      const lowerQ = question.toLowerCase();
      const hasConcept = pvStandards.concepts.some((c) => lowerQ.includes(c));
      if (!hasConcept) {
        suggestions.push(`Consider explicitly referencing: ${pvStandards.concepts.join(", ")}`);
      }
      break;
    }

    case "word_problems": {
      const wpStandards = YEAR_3_STANDARDS.word_problems;

      // Check step count (approximate by counting operations)
      if (operations.length > wpStandards.max_steps) {
        issues.push(`Multi-step problem detected (${operations.length} steps). Year 3 should focus on 1-2 step problems`);
        difficultyLevel = "challenge";
      }

      // Check if operation is appropriate
      if (operations.length === 0) {
        issues.push("No mathematical operation detected in word problem");
        topicAppropriate = false;
      }

      // Vocabulary check
      if (question.length < 20) {
        suggestions.push("Consider adding more context to make the problem relatable");
      }
      break;
    }
  }

  // General checks
  if (question.length > 200) {
    suggestions.push("Consider simplifying - long questions may be difficult for Year 3 readers");
  }

  // Check for clear question format
  if (!question.includes("?") && !question.toLowerCase().includes("calculate") && !question.toLowerCase().includes("find")) {
    suggestions.push("Consider adding a clear question or instruction");
  }

  return {
    is_valid: issues.length === 0,
    issues,
    suggestions,
    curriculum_alignment: {
      topic_appropriate: topicAppropriate,
      difficulty_level: difficultyLevel,
      year_group_match: yearGroupMatch,
    },
  };
}

File: mcp-tools/curriculum_validator.py
"""
UK Year 3 Curriculum Validator for Adaptive Maths Tutor
Validates questions against National Curriculum standards.
"""

from typing import Dict, List, Any, Optional
import re

TOOL_METADATA = {
    "name": "validate_question",
    "description": "Validates a maths question against UK Year 3 National Curriculum standards. Checks topic appropriateness, number ranges, operation complexity, and vocabulary level.",
    "parameters": {
        "question": {
            "type": "object",
            "description": "The question to validate",
            "properties": {
                "text": {"type": "string", "description": "The question text"},
                "answer": {"type": "number", "description": "The correct answer"},
                "operation": {"type": "string", "description": "The mathematical operation involved"},
                "numbers_used": {"type": "array", "description": "Numbers appearing in the question"}
            },
            "required": ["text", "answer"]
        },
        "topic": {
            "type": "string",
            "description": "The curriculum topic (e.g., 'multiplication', 'division', 'word_problems')"
        }
    },
    "returns": {
        "is_valid": {
            "type": "boolean",
            "description": "Whether the question meets Year 3 standards"
        },
        "issues": {
            "type": "array",
            "description": "List of validation issues found"
        },
        "suggestions": {
            "type": "array",
            "description": "Suggestions for improving the question"
        }
    }
}

# UK Year 3 National Curriculum Standards
YEAR_3_STANDARDS = {
    "multiplication": {
        "tables": list(range(2, 11)),  # 2-10 times tables
        "max_product": 100,
        "description": "Recall and use multiplication facts for 2, 3, 4, 5, 8, 10 tables"
    },
    "division": {
        "max_dividend": 100,
        "divisors": [2, 3, 4, 5, 8, 10],
        "description": "Divide numbers up to 100 by 2, 3, 4, 5, 8, 10"
    },
    "addition": {
        "max_sum": 1000,
        "description": "Add numbers with up to 3 digits using formal written methods"
    },
    "subtraction": {
        "max_minuend": 1000,
        "description": "Subtract numbers with up to 3 digits using formal written methods"
    },
    "place_value": {
        "max_number": 1000,
        "description": "Recognise place value of each digit in 3-digit numbers"
    },
    "word_problems": {
        "steps": 1,  # One-step problems
        "operations": ["addition", "subtraction", "multiplication", "division"],
        "description": "Solve one-step problems involving all four operations"
    },
    "fractions": {
        "denominators": [2, 3, 4, 5, 8, 10],
        "description": "Recognise and use fractions with small denominators"
    }
}

# Vocabulary appropriate for Year 3 (ages 7-8)
APPROPRIATE_VOCABULARY = {
    "number_words": ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten", 
                     "eleven", "twelve", "hundred", "thousand"],
    "operation_words": ["add", "plus", "more", "total", "altogether", "sum",
                        "subtract", "minus", "take away", "less", "fewer", "difference",
                        "multiply", "times", "groups of", "lots of",
                        "divide", "share", "split", "equal groups"],
    "context_words": ["apples", "toys", "books", "children", "sweets", "pencils", "coins", 
                      "stickers", "marbles", "cakes", "biscuits", "flowers", "balls"]
}

# Words that may be too complex for Year 3
COMPLEX_VOCABULARY = [
    "calculate", "determine", "evaluate", "compute", "quotient", "dividend", 
    "multiplicand", "minuend", "subtrahend", "addend", "augend", "integer",
    "subsequently", "therefore", "hence", "consequently", "approximately"
]


def validate_question(question: Dict[str, Any], topic: str) -> Dict[str, Any]:
    """
    Validate a question against UK Year 3 curriculum standards.
    
    Args:
        question: Dictionary containing:
            - text: The question text
            - answer: The correct answer
            - operation: The mathematical operation (optional)
            - numbers_used: Numbers in the question (optional)
        topic: The curriculum topic being assessed
    
    Returns:
        Dictionary containing:
            - is_valid: Boolean indicating if question meets standards
            - issues: List of problems found
            - suggestions: List of improvement suggestions
    """
    issues = []
    suggestions = []
    
    text = question.get("text", "")
    answer = question.get("answer")
    numbers_used = question.get("numbers_used", _extract_numbers(text))
    operation = question.get("operation", _infer_operation(text))
    
    # Validate based on topic
    if topic in YEAR_3_STANDARDS:
        standard = YEAR_3_STANDARDS[topic]
        topic_issues, topic_suggestions = _validate_topic_standards(
            topic, numbers_used, answer, operation, standard
        )
        issues.extend(topic_issues)
        suggestions.extend(topic_suggestions)
    else:
        suggestions.append(f"Topic '{topic}' not in standard curriculum list. Consider using: {list(YEAR_3_STANDARDS.keys())}")
    
    # Validate vocabulary
    vocab_issues, vocab_suggestions = _validate_vocabulary(text)
    issues.extend(vocab_issues)
    suggestions.extend(vocab_suggestions)
    
    # Validate question structure
    structure_issues, structure_suggestions = _validate_structure(text, topic)
    issues.extend(structure_issues)
    suggestions.extend(structure_suggestions)
    
    is_valid = len(issues) == 0
    
    return {
        "is_valid": is_valid,
        "issues": issues,
        "suggestions": suggestions
    }


def _extract_numbers(text: str) -> List[int]:
    """Extract all numbers from question text."""
    numbers = re.findall(r'\b\d+\b', text)
    return [int(n) for n in numbers]


def _infer_operation(text: str) -> Optional[str]:
    """Infer the mathematical operation from question text."""
    text_lower = text.lower()
    
    if any(word in text_lower for word in ["times", "multiply", "groups of", "lots of"]):
        return "multiplication"
    elif any(word in text_lower for word in ["divide", "share", "split", "equal groups"]):
        return "division"
    elif any(word in text_lower for word in ["add", "plus", "more", "total", "altogether"]):
        return "addition"
    elif any(word in text_lower for word in ["subtract", "minus", "take away", "less", "fewer"]):
        return "subtraction"
    
    return None


def _validate_topic_standards(
    topic: str, 
    numbers: List[int], 
    answer: Any,
    operation: Optional[str],
    standard: Dict[str, Any]
) -> tuple[List[str], List[str]]:
    """Validate against topic-specific standards."""
    issues = []
    suggestions = []
    
    if topic == "multiplication":
        # Check times tables range
        for num in numbers:
            if num > 10 and num not in [n * m for n in range(2, 11) for m in range(1, 13)]:
                issues.append(f"Number {num} may be outside Year 3 multiplication range")
        
        if answer and answer > standard["max_product"]:
            issues.append(f"Product {answer} exceeds Year 3 maximum of {standard['max_product']}")
            suggestions.append("Consider using smaller factors to keep product under 100")
    
    elif topic == "division":
        if numbers:
            dividend = max(numbers)
            if dividend > standard["max_dividend"]:
                issues.append(f"Dividend {dividend} exceeds Year 3 maximum of {standard['max_dividend']}")
        
        # Check divisors
        for num in numbers:
            if num in [2, 3, 4, 5, 8, 10]:
                continue  # Valid divisor
            if num > 10:
                continue  # Likely the dividend
            if num not in standard["divisors"]:
                suggestions.append(f"Consider using divisors from {standard['divisors']}")
    
    elif topic == "addition":
        if answer and answer > standard["max_sum"]:
            issues.append(f"Sum {answer} exceeds Year 3 maximum of {standard['max_sum']}")
    
    elif topic == "subtraction":
        if numbers and max(numbers) > standard["max_minuend"]:
            issues.append(f"Number {max(numbers)} exceeds Year 3 maximum of {standard['max_minuend']}")
    
    elif topic == "word_problems":
        # Check for multi-step problems
        operation_count = sum(1 for op in ["add", "subtract", "multiply", "divide", "plus", "minus", "times"]
                             if op in str(numbers).lower())
        if operation_count > 1:
            issues.append("Year 3 word problems should be one-step only")
            suggestions.append("Simplify to a single operation")
    
    elif topic == "place_value":
        for num in numbers:
            if num > standard["max_number"]:
                issues.append(f"Number {num} exceeds Year 3 place value range of {standard['max_number']}")
    
    return issues, suggestions


def _validate_vocabulary(text: str) -> tuple[List[str], List[str]]:
    """Check vocabulary is age-appropriate."""
    issues = []
    suggestions = []
    
    text_lower = text.lower()
    
    # Check for complex vocabulary
    for word in COMPLEX_VOCABULARY:
        if word in text_lower:
            issues.append(f"Vocabulary '{word}' may be too complex for Year 3")
            suggestions.append(f"Consider simpler alternatives for '{word}'")
    
    # Check sentence length (Year 3 should have shorter sentences)
    sentences = text.split('.')
    for sentence in sentences:
        word_count = len(sentence.split())
        if word_count > 20:
            suggestions.append("Consider shorter sentences (under 20 words) for Year 3")
            break
    
    return issues, suggestions


def _validate_structure(text: str, topic: str) -> tuple[List[str], List[str]]:
    """Validate question structure."""
    issues = []
    suggestions = []
    
    # Check for question mark
    if '?' not in text:
        suggestions.append("Consider ending with a question mark for clarity")
    
    # Check minimum length
    if len(text) < 10:
        issues.append("Question text too short - may lack context")
    
    # For word problems, check for context
    if topic == "word_problems":
        has_context = any(word in text.lower() for word in APPROPRIATE_VOCABULARY["context_words"])
        if not has_context:
            suggestions.append("Word problems should include real-world context (e.g., apples, toys, books)")
    
    return issues, suggestions


# Example usage
if __name__ == "__main__":
    # Valid question
    valid_q = {
        "text": "Sam has 24 stickers. He shares them equally among 4 friends. How many stickers does each friend get?",
        "answer": 6,
        "operation": "division"
    }
    
    result = validate_question(valid_q, "division")
    print("Valid Question Test:")
    print(f"  Is Valid: {result['is_valid']}")
    print(f"  Issues: {result['issues']}")
    print(f"  Suggestions: {result['suggestions']}")
    
    print()
    
    # Invalid question (numbers too large)
    invalid_q = {
        "text": "Calculate the quotient when 500 is divided by 25.",
        "answer": 20,
        "operation": "division"
    }
    
    result = validate_question(invalid_q, "division")
    print("Invalid Question Test:")
    print(f"  Is Valid: {result['is_valid']}")
    print(f"  Issues: {result['issues']}")
    print(f"  Suggestions: {result['suggestions']}")
